File: app/src/main/java/team58/cs2340/donationtracker/Controllers/SearchActivity.java
package team58.cs2340.donationtracker.Controllers;

import android.content.Intent;
import android.support.annotation.NonNull;
import android.support.v7.app.AppCompatActivity;
import android.os.Bundle;
import android.util.Log;
import android.view.View;
import android.widget.AdapterView;
import android.widget.ArrayAdapter;
import android.widget.ListView;
import android.widget.Spinner;
import android.widget.TextView;
import android.widget.Toast;

import com.google.android.gms.tasks.OnCompleteListener;
import com.google.android.gms.tasks.Task;
import com.google.firebase.firestore.DocumentSnapshot;
import com.google.firebase.firestore.FirebaseFirestore;
import com.google.firebase.firestore.QuerySnapshot;

import java.util.ArrayList;
import java.util.List;

import team58.cs2340.donationtracker.Models.Category;
import team58.cs2340.donationtracker.Models.Donation;
import team58.cs2340.donationtracker.Models.DonationManager;
import team58.cs2340.donationtracker.Models.Location;
import team58.cs2340.donationtracker.Models.LocationManager;
import team58.cs2340.donationtracker.R;

public class SearchActivity extends AppCompatActivity {


    private LocationManager locationManager;
    private DonationManager donationManager;

    private TextView name;
    private Spinner locationSpinner;
    private List<Location> locations;
    private Spinner categorySpinner;
    private ListView donationList;
    private TextView message;

    private ArrayList<Donation> result;
    FirebaseFirestore db;

    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        setContentView(R.layout.activity_search);

        this.locationManager = LocationManager.getInstance();
        this.donationManager = DonationManager.getInstance();
        db = FirebaseFirestore.getInstance();

        this.name = findViewById(R.id.name);
        this.locationSpinner = findViewById(R.id.locationSpinner);
        this.categorySpinner = findViewById(R.id.categorySpinner);

        locations = new ArrayList<>();
        locations.add(locationManager.getDefaultAllLocation());
        locations.addAll(locationManager.getLocations());
        ArrayAdapter<Location> locationAdapter = new ArrayAdapter<>(this, android.R.layout.simple_spinner_item, locations);
        locationSpinner.setAdapter(locationAdapter);

        ArrayAdapter<Category> categoryArrayAdapter = new ArrayAdapter<>(this,android.R.layout.simple_spinner_item, Category.values());
        categoryArrayAdapter.setDropDownViewResource(android.R.layout.simple_spinner_dropdown_item);
        categorySpinner.setAdapter(categoryArrayAdapter);

        donationList = findViewById(R.id.donationList);
        result = new ArrayList<>();

        DonationListAdapter donationAdapter = new DonationListAdapter(this, R.layout.layout_donationitem, result);
        donationList.setAdapter(donationAdapter);

        donationList.setOnItemClickListener(new AdapterView.OnItemClickListener() {
            @Override
            public void onItemClick(AdapterView<?> parent, View view, int position, long id) {
                Intent intent = new Intent(SearchActivity.this, DonationItemDetail.class);
                intent.putExtra("donation", result.get(position));
                startActivity(intent);
            }
        });

        this.message = findViewById(R.id.message);
        message.setVisibility(View.GONE);
    }

    public void onSearchByCategory(View view) {
        message.setVisibility(View.GONE);
        result = donationManager.searchByCategory((Location) locationSpinner.getSelectedItem(),
                (Category) categorySpinner.getSelectedItem());
        DonationListAdapter donationAdapter = new DonationListAdapter(this, R.layout.layout_donationitem, result);
        donationList.setAdapter(donationAdapter);

        db.collection("donations")
                .whereEqualTo("location", ((Location) locationSpinner.getSelectedItem()).getName())
                .whereEqualTo("category", ((Category) categorySpinner.getSelectedItem()).toString())
                .get()
                .addOnCompleteListener(new OnCompleteListener<QuerySnapshot>() {
                    @Override
                    public void onComplete(@NonNull Task<QuerySnapshot> task) {
                        if (task.isSuccessful()) {
                            for (DocumentSnapshot d : task.getResult()) {
                                Donation donation = new Donation(d.getDate("timestamp"), d.getString("name"),
                                        d.getString("location"), Double.parseDouble(d.getString("value")), d.getString("shortDescription"),
                                        d.getString("fullDescription"), Category.fromString(d.getString("category")), d.getString("comment"));
                                result.add(donation);
                                DonationListAdapter donationAdapter = new DonationListAdapter(SearchActivity.this, R.layout.layout_donationitem, result);
                                donationList.setAdapter(donationAdapter);
                            }
                        } else {
                            Toast.makeText(SearchActivity.this, task.getException().getMessage(),
                                    Toast.LENGTH_SHORT).show();
                        }
                    }
                });

        if (result.size() == 0) {
            message.setVisibility(View.VISIBLE);
        }
    }

    public void onSearchByName(View view) {
        message.setVisibility(View.GONE);
        db.collection("donations")
                .whereEqualTo("location", ((Location) locationSpinner.getSelectedItem()).getName())
                .whereEqualTo("name", name.getText().toString())
                .get()
                .addOnCompleteListener(new OnCompleteListener<QuerySnapshot>() {
                    @Override
                    public void onComplete(@NonNull Task<QuerySnapshot> task) {
                        if (task.isSuccessful()) {
                            for (DocumentSnapshot d : task.getResult()) {
                                Donation donation = new Donation(d.getDate("timestamp"), d.getString("name"),
                                        d.getString("location"), Double.parseDouble(d.getString("value")), d.getString("shortDescription"),
                                        d.getString("fullDescription"), Category.fromString(d.getString("category")), d.getString("comment"));
                                result.add(donation);
                                DonationListAdapter donationAdapter = new DonationListAdapter(SearchActivity.this, R.layout.layout_donationitem, result);
                                donationList.setAdapter(donationAdapter);
                            }
                        } else {
                            Toast.makeText(SearchActivity.this, task.getException().getMessage(),
                                    Toast.LENGTH_SHORT).show();
                        }
                    }
                });
        if (result.size() == 0) {
            message.setVisibility(View.VISIBLE);
        }
    }
}

File: app/src/main/java/team58/cs2340/donationtracker/controllers/SearchActivity.kt
package team58.cs2340.donationtracker.controllers

import android.content.Intent
import android.support.v7.app.AppCompatActivity
import android.os.Bundle
import android.view.View
import android.widget.AdapterView
import android.widget.ArrayAdapter
import android.widget.EditText
import android.widget.ListAdapter
import android.widget.ListView
import android.widget.Spinner
import android.widget.TextView
import android.widget.Toast

import com.google.android.gms.tasks.OnCompleteListener
import com.google.android.gms.tasks.Task
import com.google.firebase.firestore.FirebaseFirestore
import com.google.firebase.firestore.QueryDocumentSnapshot
import com.google.firebase.firestore.QuerySnapshot

import me.xdrop.fuzzywuzzy.FuzzySearch
import team58.cs2340.donationtracker.models.Category
import team58.cs2340.donationtracker.models.Donation
import team58.cs2340.donationtracker.models.Location
import team58.cs2340.donationtracker.models.LocationsLocal
import team58.cs2340.donationtracker.R
import java.util.*

/**
 * Class for search activity
 */
class SearchActivity : AppCompatActivity() {

    private var locationSpinner: Spinner? = null
    private var categorySpinner: Spinner? = null
    private var searchTxt: EditText? = null
    private var donationList: ListView? = null
    private var nfMessageText: TextView? = null

    private var db: FirebaseFirestore? = null

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        setContentView(R.layout.activity_search_donations)

        locationSpinner = findViewById(R.id.locationSpinner)
        categorySpinner = findViewById(R.id.categorySpinner)
        searchTxt = findViewById(R.id.searchText)
        donationList = findViewById(R.id.donationSearchList)
        nfMessageText = findViewById(R.id.nothingFoundText)
        nfMessageText!!.visibility = View.GONE

        db = FirebaseFirestore.getInstance()
        val locationManager = LocationsLocal.getInstance()
        val locations = ArrayList<Location>()
        locations.add(LocationsLocal.getDefaultAllLocation())
        locations.addAll(locationManager.locations)
        val categories = ArrayList<String>()
        categories.add("All Categories")
        for (c in Category.values()) {
            categories.add(c.toString())
        }

        val locationAdapter = ArrayAdapter(this,
                android.R.layout.simple_spinner_item, locations)
        locationSpinner!!.adapter = locationAdapter
        val categoryAdapter = ArrayAdapter(this,
                android.R.layout.simple_spinner_item, categories)
        categorySpinner!!.adapter = categoryAdapter

    }

    /**
     * Function called when search button is clicked
     * @param view view of the search screen
     */
    fun onSearchClick(view: View) {
        nfMessageText!!.visibility = View.GONE
        val location = (locationSpinner!!.selectedItem as Location).name
        val category = categorySpinner!!.selectedItem as String
        val searchQuery = searchTxt!!.text.toString()

        //final ArrayList<Donation> donationsFiltered = new ArrayList<>();
        //final DonationNameToStringFunc dnFunc = new DonationNameToStringFunc();
        val donationsQueryResults = ArrayList<Donation>()

        if ("All Locations" == location && "All Categories" == category) {
            db!!.collection("donations")
                    .get()
                    .addOnCompleteListener { task ->
                        if (task.isSuccessful) {
                            for (d in Objects.requireNonNull<QuerySnapshot>(
                                    task.result)) {
                                if (FuzzySearch.weightedRatio(searchQuery,
                                                d.getString("name")) >= THRESHOLD_RATIO) {
                                    val donation = Donation(Objects.requireNonNull<Date>(
                                            d.getDate("timestamp")),
                                            d.getString("name"),
                                            d.getString("location"),
                                            java.lang.Double.parseDouble(d.getString("value")),
                                            d.getString("shortDescription"),
                                            d.getString("fullDescription"),
                                            Category.fromString(d.getString("category")!!),
                                            d.getString("comment"))
                                    donationsQueryResults.add(donation)
                                }

                                /*Donation donation = new Donation(d.getDate("timestamp"),
                                     d.getString("name"),
                                            d.getString("location"),
                                             Double.parseDouble(d.getString("value")),
                                             d.getString("shortDescription"),
                                            d.getString("fullDescription"),
                                            Category.fromString(d.getString("category")),
                                            d.getString("comment"));
                                    donationsFiltered.add(donation);*/
                            }
                            /*List<BoundExtractedResult<Donation>> matches =
                                FuzzySearch.extractTop(searchQuery, donationsFiltered, dnFunc, 5);
                                for (BoundExtractedResult result : matches) {
                                    donationsQueryResults.add((Donation) result.getReferent());
                                }*/
                            if (donationsQueryResults.isEmpty()) {
                                nfMessageText!!.visibility = View.VISIBLE
                            }
                            val donationAdapter = DonationListAdapter(
                                    this@SearchActivity, donationsQueryResults)
                            donationList!!.adapter = donationAdapter

                            donationList!!.onItemClickListener = AdapterView.OnItemClickListener { parent, view, position, id ->
                                val intent = Intent(
                                        this@SearchActivity,
                                        DonationDetailsActivity::class.java)
                                intent.putExtra("donation",
                                        donationsQueryResults[position])
                                startActivity(intent)
                            }
                        } else {
                            Toast.makeText(this@SearchActivity,
                                    Objects.requireNonNull<Exception>(task.exception).message,
                                    Toast.LENGTH_SHORT).show()
                        }
                    }

        } else if ("All Locations" == location) {
            db!!.collection("donations")
                    .whereEqualTo("category", category)
                    .get()
                    .addOnCompleteListener { task ->
                        if (task.isSuccessful) {
                            for (d in Objects.requireNonNull<QuerySnapshot>(
                                    task.result)) {
                                if (FuzzySearch.weightedRatio(searchQuery,
                                                d.getString("name")) >= THRESHOLD_RATIO) {
                                    val donation = Donation(Objects.requireNonNull<Date>(
                                            d.getDate("timestamp")),
                                            d.getString("name"),
                                            d.getString("location"),
                                            java.lang.Double.parseDouble(d.getString("value")),
                                            d.getString("shortDescription"),
                                            d.getString("fullDescription"),
                                            Category.fromString(d.getString("category")!!),
                                            d.getString("comment"))
                                    donationsQueryResults.add(donation)
                                }

                                /*Donation donation = new Donation(d.getDate("timestamp"),
                                    d.getString("name"),
                                            d.getString("location"),
                                            Double.parseDouble(d.getString("value")),
                                            d.getString("shortDescription"),
                                            d.getString("fullDescription"),
                                            Category.fromString(d.getString("category")),
                                            d.getString("comment"));
                                    donationsFiltered.add(donation);*/
                            }
                            /*List<BoundExtractedResult<Donation>> matches =
                                FuzzySearch.extractTop(searchQuery, donationsFiltered, dnFunc, 5);
                                for (BoundExtractedResult result : matches) {
                                    donationsQueryResults.add((Donation) result.getReferent());
                                }*/
                            if (donationsQueryResults.isEmpty()) {
                                nfMessageText!!.visibility = View.VISIBLE
                            }
                            val donationAdapter = DonationListAdapter(
                                    this@SearchActivity, donationsQueryResults)
                            donationList!!.adapter = donationAdapter

                            donationList!!.onItemClickListener = AdapterView.OnItemClickListener { parent, view, position, id ->
                                val intent = Intent(
                                        this@SearchActivity,
                                        DonationDetailsActivity::class.java)
                                intent.putExtra("donation",
                                        donationsQueryResults[position])
                                startActivity(intent)
                            }
                        } else {
                            Toast.makeText(this@SearchActivity,
                                    Objects.requireNonNull<Exception>(task.exception).message,
                                    Toast.LENGTH_SHORT).show()
                        }
                    }

        } else if ("All Categories" == category) {
            db!!.collection("donations")
                    .whereEqualTo("location", location)
                    .get()
                    .addOnCompleteListener { task ->
                        if (task.isSuccessful) {
                            for (d in Objects.requireNonNull<QuerySnapshot>(
                                    task.result)) {
                                if (FuzzySearch.weightedRatio(searchQuery,
                                                d.getString("name")) >= THRESHOLD_RATIO) {
                                    val donation = Donation(Objects.requireNonNull<Date>(
                                            d.getDate("timestamp")),
                                            d.getString("name"),
                                            d.getString("location"),
                                            java.lang.Double.parseDouble(d.getString("value")),
                                            d.getString("shortDescription"),
                                            d.getString("fullDescription"),
                                            Category.fromString(d.getString("category")!!),
                                            d.getString("comment"))
                                    donationsQueryResults.add(donation)
                                }

                                /*Donation donation = new Donation(d.getDate("timestamp"),
                                    d.getString("name"),
                                            d.getString("location"),
                                            Double.parseDouble(d.getString("value")),
                                            d.getString("shortDescription"),
                                            d.getString("fullDescription"),
                                            Category.fromString(d.getString("category")),
                                            d.getString("comment"));
                                    donationsFiltered.add(donation);*/
                            }
                            /*List<BoundExtractedResult<Donation>> matches =
                                FuzzySearch.extractTop(searchQuery, donationsFiltered, dnFunc, 5);
                                for (BoundExtractedResult result : matches) {
                                    donationsQueryResults.add((Donation) result.getReferent());
                                }*/
                            if (donationsQueryResults.isEmpty()) {
                                nfMessageText!!.visibility = View.VISIBLE
                            }
                            val donationAdapter = DonationListAdapter(
                                    this@SearchActivity, donationsQueryResults)
                            donationList!!.adapter = donationAdapter

                            donationList!!.onItemClickListener = AdapterView.OnItemClickListener { parent, view, position, id ->
                                val intent = Intent(
                                        this@SearchActivity,
                                        DonationDetailsActivity::class.java)
                                intent.putExtra("donation",
                                        donationsQueryResults[position])
                                startActivity(intent)
                            }
                        } else {
                            Toast.makeText(this@SearchActivity,
                                    Objects.requireNonNull<Exception>(task.exception).message,
                                    Toast.LENGTH_SHORT).show()
                        }
                    }
        } else {
            db!!.collection("donations")
                    .whereEqualTo("location", location)
                    .whereEqualTo("category", category)
                    .get()
                    .addOnCompleteListener { task ->
                        if (task.isSuccessful) {
                            for (d in Objects.requireNonNull<QuerySnapshot>(
                                    task.result)) {
                                if (FuzzySearch.weightedRatio(searchQuery,
                                                d.getString("name")) >= THRESHOLD_RATIO) {
                                    val donation = Donation(
                                            Objects.requireNonNull<Date>(d.getDate(
                                                    "timestamp")),
                                            d.getString("name"),
                                            d.getString("location"),
                                            java.lang.Double.parseDouble(d.getString("value")),
                                            d.getString("shortDescription"),
                                            d.getString("fullDescription"),
                                            Category.fromString(d.getString("category")!!),
                                            d.getString("comment"))
                                    donationsQueryResults.add(donation)
                                }

                                /*Donation donation = new Donation(d.getDate("timestamp"),
                                    d.getString("name"),
                                            d.getString("location"),
                                             Double.parseDouble(d.getString("value")),
                                             d.getString("shortDescription"),
                                            d.getString("fullDescription"),
                                            Category.fromString(d.getString("category")),
                                            d.getString("comment"));
                                    donationsFiltered.add(donation);*/
                            }
                            /*List<BoundExtractedResult<Donation>> matches =
                                FuzzySearch.extractTop(searchQuery, donationsFiltered, dnFunc, 5);
                                for (BoundExtractedResult result : matches) {
                                    donationsQueryResults.add((Donation) result.getReferent());
                                }*/
                            if (donationsQueryResults.isEmpty()) {
                                nfMessageText!!.visibility = View.VISIBLE
                            }
                            val donationAdapter = DonationListAdapter(
                                    this@SearchActivity, donationsQueryResults)
                            donationList!!.adapter = donationAdapter

                            donationList!!.onItemClickListener = AdapterView.OnItemClickListener { parent, view, position, id ->
                                val intent = Intent(
                                        this@SearchActivity,
                                        DonationDetailsActivity::class.java)
                                intent.putExtra("donation",
                                        donationsQueryResults[position])
                                startActivity(intent)
                            }
                        } else {
                            Toast.makeText(this@SearchActivity,
                                    Objects.requireNonNull<Exception>(task.exception).message,
                                    Toast.LENGTH_SHORT).show()
                        }
                    }
        }
    }

    companion object {

        private val THRESHOLD_RATIO = 60
    }
    /*class DonationNameToStringFunc implements ToStringFunction<Donation> {
        @Override
        public String apply(Donation item) {
            return item.getName();
        }
    }*/
}
